public/request/user/update.php

```diff
@@ -28,13 +29,36 @@
 if ($propertyType === UserAction::UpdatePermissions) {
     $response = SetAccountPermissionsJSON($foundSourceUser->User, $permissions, $targetUsername, $value);
 
-    // auto-apply forums permissions
-    if (
-        $response['Success']
-        && $value >= Permissions::JuniorDeveloper
-        && !$foundTargetUser->ManuallyVerified
-    ) {
-        setAccountForumPostAuth($foundSourceUser, $permissions, $foundTargetUser, authorize: true);
+    if ($response['Success']) {
+        // Auto-apply forums permissions.
+        if ($value >= Permissions::JuniorDeveloper && !$foundTargetUser->ManuallyVerified) {
+            setAccountForumPostAuth($foundSourceUser, $permissions, $foundTargetUser, authorize: true);
+        }
+
+        // Adjust attached roles.
+        if ($value <= Permissions::Unregistered) {
+            $foundTargetUser->roles()->detach();
+        }
+        if ($value === Permissions::Registered) {
+            $foundTargetUser->removeRole(Role::DEVELOPER_JUNIOR);
+            $foundTargetUser->removeRole(Role::DEVELOPER);
+            $foundTargetUser->removeRole(Role::MODERATOR);
+        }
+        if ($value === Permissions::JuniorDeveloper) {
+            $foundTargetUser->removeRole(Role::DEVELOPER);
+            $foundTargetUser->removeRole(Role::MODERATOR);
+
+            $foundTargetUser->assignRole(Role::DEVELOPER_JUNIOR);
+        }
+        if ($value === Permissions::Developer) {
+            $foundTargetUser->removeRole(Role::DEVELOPER_JUNIOR);
+            $foundTargetUser->removeRole(Role::MODERATOR);
+
+            $foundTargetUser->assignRole(Role::DEVELOPER);
+        }
+        if ($value === Permissions::Moderator) {
+            $foundTargetUser->assignRole(Role::MODERATOR);
+        }
```

Comment:
This code doesn't revoke the Moderator role when a moderator is demoted, but it's currently impossible to promote or demote a moderator from the UI as a user's current rank and new rank must both be below the rank of the user making the change.

Reply:
I had forgotten about that restriction. Given that this legacy endpoint is tightly coupled to said UI, it seems like what may be best is to just exclude a condition for the `Moderator` permission altogether.